pkg/firewall/client.go

```diff
@@ -2,125 +2,147 @@ package firewall
-	fwName := bastionFirewallPolicyRuleName(cluster.Name)
-	ipProtocol := ProtocolTCP
-
-	tagName := fmt.Sprintf("%s-bastion", cluster.GetName())
-	logger := c.logger.WithValues("name", tagName)
-	logger.Info("Creating firewall rule for bastion")
-
-	rule := &computepb.Firewall{
-		Allowed: []*computepb.Allowed{
-			{
-				IPProtocol: &ipProtocol,
-				Ports:      []string{PortSSH},
-			},
-		},
-		Description: proto.String("allow port 22 for SSH to"),
-		Direction:   proto.String(computepb.Firewall_INGRESS.String()),
-		Name:        &fwName,
-		Network:     cluster.Status.Network.SelfLink,
-		TargetTags:  []string{tagName},
-	}
+	logger.Info("Creating firewall rule")
+	defer logger.Info("Done creating firewall rule")
+
+	firewall := toGCPFirewall(cluster, rule)
 
 	req := &computepb.InsertFirewallRequest{
 		Project:          cluster.Spec.Project,
-		FirewallResource: rule,
+		FirewallResource: firewall,
 	}
+	op, err := c.firewallClient.Insert(ctx, req)
 
-	op, err := c.fwService.Insert(ctx, req)
-	if err != nil {
-		if isAlreadyExistError(err) {
-			// pass thru, resource already exists
-			return nil
-		} else {
-			return errors.WithStack(err)
-		}
+	if hasHttpCode(err, http.StatusConflict) {
+		logger.Info("Firewall already exists. Updating")
+		err = c.updateFirewall(ctx, cluster, firewall)
+		return errors.WithStack(err)
 	}
 
-	err = op.Wait(ctx)
-
 	if err != nil {
-		if isAlreadyExistError(err) {
-			// pass thru, resource already exists
-		} else {
-			return errors.WithStack(err)
-		}
+		return errors.WithStack(err)
 	}
 
-	logger.Info("Created firewall rule for bastion")
-	return nil
+	err = op.Wait(ctx)
+	return errors.WithStack(err)
 }
 
-func (c *Client) DeleteBastionFirewallRule(ctx context.Context, cluster *capg.GCPCluster) error {
-	name := fmt.Sprintf("%s-bastion", cluster.GetName())
-	logger := c.logger.WithValues("name", name)
-	logger.Info("Deleting firewall rule for bastion")
+func (c *Client) DeleteRule(ctx context.Context, cluster *capg.GCPCluster, ruleName string) error {
+	logger := c.getLogger(ctx, ruleName)
+
+	logger.Info("Deleting firewall rule")
+	defer logger.Info("Done deleting firewall rule")
 
 	req := &computepb.DeleteFirewallRequest{
 		Project:  cluster.Spec.Project,
-		Firewall: bastionFirewallPolicyRuleName(cluster.Name),
+		Firewall: ruleName,
 	}
-
-	op, err := c.fwService.Delete(ctx, req)
-
-	if isNotFoundError(err) {
-		// pass thru, resource is already deleted
-		logger.Info("Firewall rule for bastion is already deleted")
-
+	op, err := c.firewallClient.Delete(ctx, req)
+	if hasHttpCode(err, http.StatusNotFound) {
+		logger.Info("Firewall already deleted")
 		return nil
-	} else if err != nil {
-		return errors.WithStack(err)
 	}
+
 	err = op.Wait(ctx)
 
-	if isNotFoundError(err) {
-		// pass thru, resource is already deleted
-	} else if err != nil {
+	return errors.WithStack(err)
+}
+
+func (c *Client) updateFirewall(ctx context.Context, cluster *capg.GCPCluster, firewall *computepb.Firewall) error {
```

Comment:
i personally don't like that the patch will occur every time even if there is no change, that can be quite heavy on write API calls and in big scale there might be some issues with API limits, we saw something like this on AWS, no sure how google handle such situation, maybe its ok 

I would rather try check the rule and compare  some important values like IPs and patch only if they differ

Reply:
I think it's a legit concern. I just did it this way because I thought it would be simpler.
It looks like GCP has different rate limiting for GET and PATCH operations. [According to docs]() the firewall operations fall under `Read requests - Limit for *.get methods: 25 requests/second (maximum of 1500 requests/minute)` for GET and `Heavy-weight mutation requests - Limit for patch, delete, and insert methods for the interconnects and interconnectAttachments resources: 12.5 requests/second (maximum of 750 requests/minute)`.

Mind you this is per project so this means all operations we do across all operators. So it is better to do a GET first and only do the PATCH if there are changes. I'll push a change

Reply:
sounds good, thanks for investigating

Reply:
I'll merge it in like that and if you have any concerns about it we can fix it in the follow up story